spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/FilterPushdown.scala

```diff
@@ -0,0 +1,158 @@
+package org.apache.spark.sql.cassandra
+
+import com.datastax.spark.connector.cql.TableDef
+import org.apache.spark.sql.sources
+import org.apache.spark.sql.sources.Filter
+
+
+/**
+ * Calculate the pushdown predicates for a given table and predicates.
+ *
+ * Partition pruning predicates are also detected an applied.
+ *  1. Only push down no-partition key column predicates with =, >, <, >=, <= predicate
+ *  2. Only push down primary key column predicates with = or IN predicate.
+ *  3. If there are regular columns in the pushdown predicates, they should have
+ *     at least one EQ expression on an indexed column and no IN predicates.
+ *  4. All partition column predicates must be included in the predicates to be pushed down,
+ *     only the last part of the partition key can be an IN predicate. For each partition column,
+ *     only one predicate is allowed.
+ *  5. For cluster column predicates, only last predicate can be non-EQ predicate
+ *     including IN predicate, and preceding column predicates must be EQ predicates.
+ *     If there is only one cluster column predicate, the predicates could be any non-IN predicate.
+ *  6. There is no pushdown predicates if there is any OR condition or NOT IN condition.
+ *  7. We're not allowed to push down multiple predicates for the same column if any of them
+ *     is equality or IN predicate.
+ *
+ */
+object FilterPushdown {
```

Comment:
A déjà vu :) Now I get what you're trying to say back then that you had to redo the Catalog, too. 

So, we have a problem that previously our predicate pushdown code worked on `Expression`s. But new Spark introduced `Filter`, which is a special kind of an `Expression` for filtering, yet it doesn't implement the internal `Expression` class. So `Filter` and `Expression` are two classes serving almost the same purpose for us, but they are separate and have no common interface. Let's discuss how to get out of this trap.

Proposed solutions:
  1. Forget Expressions. Rewrite everything to use Filters. This way we get elegant code, possibly less dependent on internal Spark APIs (which might break any time), but we're not finished until we fix all the other code and the size of the changeset gets through the roof.
  1. Use old-school engineering. Copy-paste the whole thing, then find-and-replace `Expression` with `Filter`, slightly fix a few places where it doesn't compile and we're done!  BTW: This is the current solution here.
  1. Write *common* code that works both for `Expressions` and `Filters`. DRY magic.

@jacek-lewandowski @RussellSpitzer @helena @pedjak @bcantoni which one do you like the most?
The voting starts here; you can of course add more solutions.